src/PyProxyHandler.cc

```diff
@@ -176,29 +176,806 @@ bool PyBaseProxyHandler::isExtensible(JSContext *cx, JS::HandleObject proxy,
+      PyList_SetItem(self, actualTarget, itemStart);
+
+      actualStart++;
+      actualTarget++;
+      count--;
+    }
+  }
+
+  // return ref to self
+  args.rval().set(jsTypeFactory(cx, self));
+  return true;
+}
+
+// private util
+static bool array_copy_func(JSContext *cx, unsigned argc, JS::Value *vp, const char *fName, bool checkRequireAtLeastOne = true) {
+  JS::CallArgs args = JS::CallArgsFromVp(argc, vp);
+
+  if (checkRequireAtLeastOne && !args.requireAtLeast(cx, fName, 1)) {
+    return false;
+  }
+
+  JS::RootedObject proxy(cx, JS::ToObject(cx, args.thisv()));
+  if (!proxy) {
+    return false;
+  }
+  PyObject *self = JS::GetMaybePtrFromReservedSlot<PyObject>(proxy, PyObjectSlot);
+
+  JS::RootedObject selfCopy(cx, &jsTypeFactoryCopy(cx, self).toObject());
+
+  JS::HandleValueArray jArgs(args);
+  JS::RootedValue rval(cx);
+
+  if (!JS_CallFunctionName(cx, selfCopy, fName, jArgs, &rval)) {
+    return false;
+  }
+
+  args.rval().set(rval);
+
+  return true;
+}
+
+static bool array_includes(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "includes");
+}
+
+static bool array_forEach(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "forEach");
+}
+
+static bool array_map(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "map");
+}
+
+static bool array_filter(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "filter");
+}
+
+static bool array_reduce(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "reduce");
+}
+
+static bool array_reduceRight(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "reduceRight");
+}
+
+static bool array_some(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "some");
+}
+
+static bool array_every(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "every");
+}
+
+static bool array_find(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "find");
+}
+
+static bool array_findIndex(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "findIndex");
+}
+
+static bool array_flat(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "flat", false);
+}
+
+static bool array_flatMap(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "flatMap");
+}
+
+static bool array_join(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "join", false);
+}
+
+static bool array_toString(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "toString", false);
+}
+
+static bool array_toLocaleString(JSContext *cx, unsigned argc, JS::Value *vp) {
+  return array_copy_func(cx, argc, vp, "toLocaleString", false);
```

Comment:
These create an unnecessary copy of the proxy object. Also, I'm fairly confident that for the methods that take a callback function (like `.map()`), if the array parameter to the callback function is mutated, the proxy object will not reflect that mutation.

Reply:
This does applies indeed for methods that can mutate the array, will fix

Reply:
done